parser/shared/src/main/scala/jawn/AsyncParser.scala

```diff
@@ -31,6 +31,7 @@ object AsyncParser {
 
   sealed abstract class Mode(val start: Int, val value: Int)
   case object UnwrapArray extends Mode(-5, 1)
+  case object UnwrapMultiArray extends Mode(-5, 1)
   case object ValueStream extends Mode(-1, 0)
   case object SingleValue extends Mode(-1, -1)
```

Comment:
Unfortunately this change is not backwards compatible, since it adds a new member to a sealed hierarchy. This will break any previous total pattern matches since they will not be accounting for this new case.

See:
- https://github.com/lightbend/mima/issues/200

Reply:
So what is your recommendation?

Reply:
Indeed it is not a linking error, but it can still crash at runtime. If I wrote a pattern match that checks for `UnwrapArray`, `ValueStream`, and `SingleValue`, and assumes those are the only 3 possible cases (as promised by `sealed`) then that code is in big trouble if it encounters an instance of `UnwrapMultiArray`.

My recommendation is to find a way to make this change backwards-compatibly :)

Reply:
So the only way to make this binary compatible is the initial implementation which was actually changed because design wise its worse, i.e.

```scala
  sealed abstract class Mode(val start: Int, val value: Int)
  case object UnwrapArray extends Mode(-5, 1)
  case object ValueStream extends Mode(-1, 0)
  case object SingleValue extends Mode(-1, -1)

  def apply[J](mode: Mode = SingleValue): AsyncParser[J] =
    new AsyncParser(
      state = mode.start,
      curr = 0,
      context = null,
      stack = Nil,
      data = new Array[Byte](131072),
      len = 0,
      allocated = 131072,
      offset = 0,
      done = false,
      streamMode = mode.value,
      multiValue = false
    )

  def apply[J](mode: Mode, multiValue: Boolean): AsyncParser[J] =
    new AsyncParser(
      state = mode.start,
      curr = 0,
      context = null,
      stack = Nil,
      data = new Array[Byte](131072),
      len = 0,
      allocated = 131072,
      offset = 0,
      done = false,
      streamMode = mode.value,
      multiValue = multiValue
    )
}
```

The reason why design wise this is worse is because `multiValue` has no meaning for other `Mode`'s that is not `ValueStream` so its misleading. Furthermore due to how overloading works you cannot have a duplicate default parameter for `mode: Mode` (which is why the second new `apply` is missing the `= SingleValue` default param.

There was also another idea which is the make `ValueStream` a `case class` that takes a `multi-value` parameter but afaik that is not binary compatible (in the traditional/correct sense).